src/renderer/Metal.zig

```diff
@@ -581,2768 +114,236 @@ pub fn init(alloc: Allocator, options: renderer.Options) !Metal {
-            .min_contrast = options.config.min_contrast,
-            .cursor_pos = .{ std.math.maxInt(u16), std.math.maxInt(u16) },
-            .cursor_color = undefined,
-            .bg_color = .{
-                options.config.background.r,
-                options.config.background.g,
-                options.config.background.b,
-                @intFromFloat(@round(options.config.background_opacity * 255.0)),
-            },
-            .cursor_wide = false,
-            .use_display_p3 = options.config.colorspace == .@"display-p3",
-            .use_linear_blending = options.config.blending.isLinear(),
-            .use_linear_correction = options.config.blending == .@"linear-corrected",
-        },
+    // Ensure that if our layer is oversized it
+    // does not overflow the bounds of the view.
+    info.view.setProperty("clipsToBounds", true);
 
-        // Fonts
-        .font_grid = options.font_grid,
-        .font_shaper = font_shaper,
-        .font_shaper_cache = font.ShaperCache.init(),
+    // Ensure that our layer has a content scale set to
+    // match the scale factor of the window. This avoids
+    // magnification issues leading to blurry rendering.
+    layer.layer.setProperty("contentsScale", info.scaleFactor);
 
-        // Shaders (initialized below)
-        .shaders = undefined,
+    // This makes it so that our display callback will actually be called.
+    layer.layer.setProperty("needsDisplayOnBoundsChange", true);
 
-        // Metal stuff
+    return .{
         .layer = layer,
-        .display_link = display_link,
-        .terminal_colorspace = terminal_colorspace,
-        .custom_shader_state = null,
-        .gpu_state = gpu_state,
+        .device = device,
+        .queue = queue,
+        .blending = opts.config.blending,
+        .default_storage_mode = default_storage_mode,
     };
-
-    try result.initShaders();
-
-    // Do an initialize screen size setup to ensure our undefined values
-    // above are initialized.
-    try result.setScreenSize(result.size);
-
-    return result;
 }
 
 pub fn deinit(self: *Metal) void {
-    self.gpu_state.deinit();
-
-    if (DisplayLink != void) {
-        if (self.display_link) |display_link| {
-            display_link.stop() catch {};
-            display_link.release();
-        }
-    }
-
-    self.terminal_colorspace.release();
-
-    self.cells.deinit(self.alloc);
-
-    self.font_shaper.deinit();
-    self.font_shaper_cache.deinit(self.alloc);
-
-    self.config.deinit();
-
-    {
-        var it = self.images.iterator();
-        while (it.next()) |kv| kv.value_ptr.image.deinit(self.alloc);
-        self.images.deinit(self.alloc);
-    }
-    self.image_placements.deinit(self.alloc);
-
-    self.deinitShaders();
+    self.queue.release();
+    self.device.release();
 
-    self.* = undefined;
+    // NOTE: We don't release the layer here because that should be taken
+    //       care of automatically when the hosting view is destroyed.
```

Comment:
Thanks for writing this comment. It makes sense that the hosting view would retain/release its `layer`, but I'm surprised that it doesn't `retain` when we set the property so that we can release our reference. I just want to double check your note here is correct or if you have more details about it.

Reply:
I was surprised by this too! But I was getting a double-free when closing the window, and I checked the memory graph in xcode to make sure that this wasn't a leak if we didn't do it and it didn't seem to be a leak afaict.

Reply:
Resolved by manually retaining and then releasing the layer to avoid the un-zig-ish implicit autorelease behavior.